### PR TITLE
M2-4928, M2-4929 Drawing and AbTrails streaming updates

### PR DESCRIPTION
**Pull request by vmkhitaryanscn**
[M2-4928](https://mindlogger.atlassian.net/jira/software/c/projects/M2/issues/M2-4928)
[M2-4929](https://mindlogger.atlassian.net/jira/software/c/projects/M2/issues/M2-4929)

Added more information to drawing and abTrails to be sent over the socket

[M2-4928]: https://mindlogger.atlassian.net/browse/M2-4928?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[M2-4929]: https://mindlogger.atlassian.net/browse/M2-4929?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/entities/abTrail/ui/AbCanvas.tsx [hunk 1]
```diff
@@ -195,24 +196,44 @@ const AbCanvas: FC<Props> = props => {
...
+  const createStreamEventPoint = (point: Point): StreamEventPoint => {
+    const [currentNodeLabel, nextNodeLabel] = getCurrentAndNextNodeLabels();
+
+    return {
+      x: (point.x * width) / 100,
+      y: (point.y * width) / 100,
+      time: Date.now(),
+      line_number: logLines?.length - 1,
+      error: StreamEventError.NotDefined,
+      correct_path: `${currentNodeLabel} ~ ${nextNodeLabel}`,
```
Comment (anq83) [inline src/entities/abTrail/ui/AbCanvas.tsx:232]: I'd suggest to move details of creation the string into mapper layer, to make stream-point type more clear with simple primitive properties.

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/entities/abTrail/ui/AbCanvas.tsx [hunk 2]
```diff
@@ -339,7 +362,12 @@ const AbCanvas: FC<Props> = props => {
       resetCurrentPath();
       setFlareGreenPointIndex({ index: getCurrentIndex() });
       onMessage(MessageType.IncorrectLine);
+
+      streamEventPoint.error = StreamEventError.OVER_WRONG_POINT;
+      streamEventPoint.actual_path = node.label;
```
Comment (anq83) [inline src/entities/abTrail/ui/AbCanvas.tsx:367]: the same - to separated function

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/entities/abTrail/lib/types/test.ts [hunk 3]
```diff
@@ -29,10 +29,20 @@ export type OnResultLog = {
...
+  OVER_UNDEFINED_POINT = 'E2',
+}
 export type StreamEventPoint = {
   x: number;
   y: number;
   time: number;
+  line_number: number;
+  error: StreamEventError;
+  correct_path: string;
+  actual_path: string;
```
[the remark this answers is not in the document]
Comment (BamMironov) [reply src/entities/abTrail/lib/types/test.ts:45]: @anq83 
That's the expected field names by the receivers of live events, I guess

[the remark this answers is not in the document]
Comment (anq83) [reply src/entities/abTrail/lib/types/test.ts:57]: correct, they can only exist in DTO

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/entities/abTrail/ui/AbCanvas.tsx [hunk 4]
```diff
@@ -274,11 +328,8 @@ const AbCanvas: FC<Props> = props => {
     reCreatePath(point);
     drawPath();
     reRender();
-    onLog({
-      x: (touchInfo.x * width) / 100,
-      y: (touchInfo.y * width) / 100,
-      time: Date.now(),
-    });
+
+    onAddPointToStream(mapStreamEventToDto(createStreamEventPoint(point)));
```
Comment (anq83) [inline src/entities/abTrail/ui/AbCanvas.tsx:332]: Can we move mapStreamEventToDto ontide of the AB-related components?

File: src/entities/abTrail/ui/AbCanvas.tsx [hunk 5]
```diff
@@ -329,6 +376,8 @@ const AbCanvas: FC<Props> = props => {
       keepPathInState();
       reCreatePath(point);
       incrementCurrentIndex();
+
+      addOverRightPointToStream(point);
```
Comment (anq83) [inline src/entities/abTrail/ui/AbCanvas.tsx:380]: maybe addOverCorrectNextPointToStream?

File: src/entities/abTrail/ui/AbCanvas.tsx [hunk 6]
```diff
@@ -321,6 +366,8 @@ const AbCanvas: FC<Props> = props => {
       });
       onMessage(MessageType.Completed);
       onComplete();
+
+      addOverRightPointToStream(point);
```
Comment (anq83) [inline src/entities/abTrail/ui/AbCanvas.tsx:370]: maybe addOverCorrectNextPointToStream or maybe addOverCorrectLastPointToStream?

**Review by anq83 — CHANGES_REQUESTED** [review]

**Review by BamMironov — APPROVED** [review]

File: src/features/pass-survey/ui/ActivityItem.tsx [hunk 7]
```diff
@@ -67,7 +72,10 @@ function ActivityItem({
 
   const [scrollEnabled, setScrollEnabled] = useState(initialScrollEnabled);
 
-  const { sendLiveEvent } = useSendEvent(streamEnabled);
+  const { sendLiveEvent } = useSendEvent(
+    type as StreamEventActivityItemType,
```
Comment (anq83) [inline src/features/pass-survey/ui/ActivityItem.tsx:76]: is it safe to use as here?

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/shared/lib/tcp/lib/streamEventMapper.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+) => {
+  switch (type) {
+    case 'AbTest':
+      return mapABTestStreamEventToDto(streamEvent as AbTestStreamEvent);
+    case 'DrawingTest':
+      return mapDrawingStreamEventToDto(streamEvent as DrawingStreamEvent);
+    case 'StabilityTracker':
+      return mapStabilityTrackerStreamEventToDto(
+        streamEvent as StabilityTrackerEvent,
+      );
```
Comment (anq83) [inline src/shared/lib/tcp/lib/streamEventMapper.ts:77]: It seems Flanker is missed

**Review by anq83 — DISMISSED** [review]

File: src/shared/lib/tcp/useSendLiveEvent.ts [hunk 9]
```diff
@@ -1,27 +1,23 @@
...
-
 export function useSendEvent(streamEnabled: boolean) {
   const { sendMessage, connected } = useTCPSocket({
     onClosed: () => {},
   });
 
   const sendLiveEvent = useCallback(
-    (data: LiveEvent) => {
+    (dataDto: LiveEventDto) => {
+      console.log(dataDto);
```
Comment (BamMironov) [inline src/shared/lib/tcp/useSendLiveEvent.ts:13]: Please remove it

**Review by BamMironov — CHANGES_REQUESTED** [review]

**Review by BamMironov — APPROVED** [review]

**Pull request merged**